x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/InsertFieldExtraction.java

```diff
@@ -102,25 +100,17 @@ private static Set<Attribute> missingAttributes(PhysicalPlan p) {
         var missing = new LinkedHashSet<Attribute>();
         var input = p.inputSet();
 
-        // For LOOKUP JOIN we only need field-extraction on left fields used to match, since the right side is always materialized
-        if (p instanceof LookupJoinExec join) {
-            join.leftFields().forEach(f -> {
-                if (input.contains(f) == false) {
-                    missing.add(f);
-                }
-            });
-            return missing;
-        }
-
-        // collect field attributes used inside expressions
-        // TODO: Rather than going over all expressions manually, this should just call .references()
-        p.forEachExpression(TypedAttribute.class, f -> {
+        // Collect field attributes referenced by this plan but not yet present in the child's output.
+        // This is also correct for LookupJoinExec, where we only need field extraction on the left fields used to match, since the right
+        // side is always materialized.
+        p.references().forEach(f -> {
```

Comment:
I think this simplifications should prevent future weirdnesses.

In the past, we couldn't declare the attributes added by `ENRICH` or `LOOKUP_` :chicken: to be field attributes, because we'd insert faulty field extractions. That's because walking _all_ of the plan's expressions and looking for field attributes is too broad - we need only check the references (i.e. the required input fields).

This simplification is only possible since the `references` method was overhauled and fixed.